operators/tokenizer/bpe_kernels.cc

```diff
@@ -0,0 +1,359 @@
+    case 0x000C:
+    case 0x000D:
+    case 0x001C:
+    case 0x001D:
+    case 0x001E:
+    case 0x001F:
+    case 0x0020:
+    case 0x0085:
+    case 0x00A0:
+    case 0x1680:
+    case 0x2000:
+    case 0x2001:
+    case 0x2002:
+    case 0x2003:
+    case 0x2004:
+    case 0x2005:
+    case 0x2006:
+    case 0x2007:
+    case 0x2008:
+    case 0x2009:
+    case 0x200A:
+    case 0x2028:
+    case 0x2029:
+    case 0x202F:
+    case 0x205F:
+    case 0x3000:
+      return true;
+  }
+  return false;
+}
+
+bool AllSpaceUstring(const ustring& str) {
+  return std::all_of(str.begin(), str.end(), [](char32_t ch) { return IsUnicodeSpace(ch); });
+}
+
+KernelBpeTokenizer::KernelBpeTokenizer(const OrtApi& api, const OrtKernelInfo& info, const BpeModelConf& conf)
+    : BaseKernel(api, info),
+      bpe_conf_(conf) {
+  std::string vocab = ort_.KernelInfoGetAttribute<std::string>(&info, "vocab");
+  if (vocab.empty()) {
+    ORTX_CXX_API_THROW("vocabulary shouldn't be empty.", ORT_INVALID_ARGUMENT);
+  }
+
+  std::string merges = ort_.KernelInfoGetAttribute<std::string>(&info, "merges");
+  if (merges.empty()) {
+    ORTX_CXX_API_THROW("merges shouldn't be empty.", ORT_INVALID_ARGUMENT);
+  }
+
+  if (!TryToGetAttribute<int64_t>("padding_length", padding_length_)) {
+    padding_length_ = -1;
+  }
+
+  if (padding_length_ != -1 && padding_length_ <= 0) {
+    ORTX_CXX_API_THROW("padding_length should be more than 0 or equal -1", ORT_INVALID_ARGUMENT);
+  }
+
+  std::stringstream vocabu_stream(vocab);
+  std::stringstream merges_stream(merges);
+  bbpe_tokenizer_ = std::make_unique<BpeModel>();
+  bbpe_tokenizer_->Load(vocabu_stream, merges_stream, conf.unk_token_, conf.GetSpecialTokens().c_str());
+
+  // TODO: need to check if the special token ids are the same as the ones in HFTokenizer
+  unk_token_id_ = bbpe_tokenizer_->GetTokenId(conf.unk_token_);
+  if (conf.bos_token_ != nullptr) {
+    bos_token_id_ = bbpe_tokenizer_->GetTokenId(conf.bos_token_);
+  }
+  if (conf.eos_token_ != nullptr) {
+    eos_token_id_ = bbpe_tokenizer_->GetTokenId(conf.eos_token_);
+  }
+  if (conf.pad_token_ != nullptr) {
+    pad_token_id_ = bbpe_tokenizer_->GetTokenId(conf.pad_token_);
+  }
+}
+
+std::vector<int64_t> KernelBpeTokenizer::Tokenize(ustring& input,
+                                                  int64_t max_length,
+                                                  bool compute_offset_mapping,
+                                                  std::list<OffsetMappingType>& offset_map) const {
+  std::vector<int64_t> res;
+  std::list<std::pair<uint32_t, uint32_t>> byte_list;
+
+  if (ModelName() == BpeModelConf::kModel_CLIP) {
+    // Merges consecutive '\s+' for CLIP
+    /*
+      text = re.sub(r"\s+", " ", text)
+      text = text.strip()
+    */
+    ustring& str = input;
+    str.erase(std::unique(str.begin(), str.end(),
+                          [](char32_t lhs, char32_t rhs) { return IsUnicodeSpace(lhs) && IsUnicodeSpace(rhs); }),
```

Comment:
this is an interesting use of the equality predicate. however, std::unique requires an equivalence relation.

from here: https://en.cppreference.com/w/cpp/algorithm/unique
"Elements are compared using the given binary predicate p. The behavior is undefined if it is not an equivalence relation."

https://en.wikipedia.org/wiki/Equivalence_relation

Reply:
Nice catch! in the case, I have to write a loop for eliminate the spaces.

Reply:
both are UnicodeSpace is an equivalence relation, isn't it?  It satisfies reflexivity, symmetry and transitivity

Reply:
I think reflexivity is not satisfied. for example, if x is not a unicode space, then `IsUnicodeSpace(x) && IsUnicodeSpace(x) == false`